src/LowerParallelTasks.cpp

```diff
@@ -0,0 +1,506 @@
+#include "LowerParallelTasks.h"
+
+#include <string>
+
+#include "Argument.h"
+#include "Closure.h"
+#include "DebugArguments.h"
+#include "ExprUsesVar.h"
+#include "IRMutator.h"
+#include "IROperator.h"
+#include "Module.h"
+#include "Param.h"
+
+namespace Halide {
+namespace Internal {
+
+// TODO(zalman): Find a better place for this code to live.
+LoweredFunc GenerateClosureIR(const std::string &name, const Closure &closure,
+                              std::vector<LoweredArgument> &args, int closure_arg_index,
+                              const Stmt &body, const Target &t) {
+    // Figure out if user_context has to be dealt with here.
+    std::string closure_arg_name = unique_name("closure_arg");
+    args[closure_arg_index] = LoweredArgument(closure_arg_name, Argument::Kind::InputScalar,
```

Comment:
I thought we needed to keep some modulus/remainder and other analysis through closures like this. But I don't see that logic in HexagonOffload.cpp... so I'm not sure. I think maybe we should add a test that checks that loads/stores are aligned when expected (if we don't have one already), and make sure those loads/stores are in a parallel loop.

I think we used to propagate this through LoweredArgument, but then when we started adding alignment metadata to loads/stores directly, we stopped doing that.

Reply:
Is the idea of the test to have an input buffer with host alignment asserted and then make sure a load against that buffer inside a closure is aligned? (E.g. by compiling to a fixed CPU target and pattern matching the asm output, or possibly doing it on LLVM IR.)

Reply:
We also need to have constraints required to make the load aligned, e.g. a 2D buffer where the stride in y is constrained to be a multiple of the alignment.

I think it would be sufficient to check the Halide IR after lowering, that the load has the correct alignment annotation. No need to check LLVM IR/assembly I don't think.

Reply:
This pass happens at the very end of lowering, so if there was something in the Halide IR, it will already be there. (The place in lowering is sort of dictated by it being moved from backend specific codegen to a lowering pass.) But if there is lookaside or schedule information, that might need to be copied onto a new LoweredArgument or similar. I think if we really want a reliable test, it will need to validate some property of the final output. A Halide IR test may also be useful.